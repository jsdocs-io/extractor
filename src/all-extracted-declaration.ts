import type {
  ExtractedClass,
  ExtractedClassConstructor,
  ExtractedClassMethod,
  ExtractedClassProperty,
} from "./extract-class";
import type { ExtractedEnum, ExtractedEnumMember } from "./extract-enum";
import type { ExtractedFunction } from "./extract-function";
import type {
  ExtractedInterface,
  ExtractedInterfaceCallSignature,
  ExtractedInterfaceConstructSignature,
  ExtractedInterfaceGetAccessor,
  ExtractedInterfaceIndexSignature,
  ExtractedInterfaceMethod,
  ExtractedInterfaceProperty,
  ExtractedInterfaceSetAccessor,
} from "./extract-interface";
import type { ExtractedNamespace } from "./extract-namespace";
import type { ExtractedTypeAlias } from "./extract-type-alias";
import type { ExtractedVariable } from "./extract-variable";

/**
`AllExtractedDeclaration` is the union of all possible declarations
that can be extracted, with some being found only in other declarations
(e.g., class method declarations are found only in a class declaration).
*/
export type AllExtractedDeclaration =
  | ExtractedVariable
  | ExtractedFunction
  | ExtractedClass
  | ExtractedClassConstructor
  | ExtractedClassProperty
  | ExtractedClassMethod
  | ExtractedInterface
  | ExtractedInterfaceProperty
  | ExtractedInterfaceMethod
  | ExtractedInterfaceConstructSignature
  | ExtractedInterfaceCallSignature
  | ExtractedInterfaceIndexSignature
  | ExtractedInterfaceGetAccessor
  | ExtractedInterfaceSetAccessor
  | ExtractedEnum
  | ExtractedEnumMember
  | ExtractedTypeAlias
  | ExtractedNamespace;

/**
`AllExtractedDeclarationKind` is the union of all discriminators
used to detect the kind of declaration.
*/
export type AllExtractedDeclarationKind = AllExtractedDeclaration["kind"];
