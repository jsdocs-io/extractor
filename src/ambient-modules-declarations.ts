import { ModuleDeclaration, Node, type Project } from "ts-morph";
import { isHidden } from "./is-hidden";
import { sourceFilePath } from "./source-file-path";

export type AmbientModulesDeclarationsReturn = {
  containerName: string;
  exportName: string;
  declaration: ModuleDeclaration;
}[];

export const ambientModulesDeclarations = (
  containerName: string,
  project: Project,
  pkgName?: string,
): AmbientModulesDeclarationsReturn => {
  const ambientModulesDeclarations = [];
  for (const symbol of project.getAmbientModules()) {
    for (const declaration of symbol.getDeclarations()) {
      if (isHidden(declaration) || !Node.isModuleDeclaration(declaration)) {
        continue;
      }
      if (pkgName && !sourceFilePath(declaration).startsWith(`/${pkgName}`)) {
        // Ignore ambient modules that are not from the analyzed package.
        continue;
      }
      const exportName = declaration.getName();
      // TODO:
      // Remove surrounding quotes and eventual spaces
      // const exportID = exportName
      //   .replace(/"|'/g, "")
      //   .replace(/\s/g, "_")
      //   .trim();
      ambientModulesDeclarations.push({
        containerName,
        exportName,
        declaration,
      });
    }
  }
  return ambientModulesDeclarations;
};
