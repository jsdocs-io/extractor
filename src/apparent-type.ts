import { TypeFormatFlags, ts, type Node } from "ts-morph";

export function apparentType(node: Node): string {
	// See:
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-667578386
	return node
		.getType()
		.getApparentType()
		.getText(
			node,
			ts.TypeFormatFlags.NoTruncation | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope,
		)
		.replace(/^Number$/, "number")
		.replace(/^Boolean$/, "boolean")
		.replace(/^String$/, "string");
}
