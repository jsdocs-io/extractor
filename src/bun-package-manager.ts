import { Effect } from "effect";
import { execa } from "execa";
import { InstallPackageError, PackageManager } from "./package-manager";

/** @internal */
export const bunPackageManager = (bunPath = "bun") =>
	PackageManager.of({
		installPackage: ({ pkg, cwd }) =>
			Effect.gen(function* (_) {
				// Run `bun add <pkg> --verbose`.
				// See https://bun.sh/docs/cli/add.
				const { stdout } = yield* _(
					Effect.tryPromise({
						try: () => execa(bunPath, ["add", pkg, "--verbose"], { cwd }),
						catch: (e) => new InstallPackageError({ cause: e }),
					}),
				);

				// With verbose output on, bun prints one line per installed package
				// (e.g., "foo@1.0.0"), including all installed dependencies.
				// These lines are between the two delimiting lines found here:
				// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
				const lines = stdout.split("\n");
				const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
				const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
				const installedPackages = lines.slice(beginHash + 1, endHash);
				return installedPackages;
			}),
	});
