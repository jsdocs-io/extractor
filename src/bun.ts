import { execa } from "execa";

/** `Bun` is a wrapper for the Bun package manager. */
export class Bun {
	#bunCmd: string;

	constructor(bunCmd = "bun") {
		this.#bunCmd = bunCmd.trim();
	}

	/**
	`add` installs the given package in the given directory and
	returns a list of all installed packages (e.g., `["foo@1.0.0"]`).

	@see {@link https://bun.sh/docs/cli/add}
	*/
	async add(pkg: string, cwd: string): Promise<string[]> {
		// Run `bun add <pkg> --verbose` in the `cwd`.
		const { stdout } = await execa(this.#bunCmd, ["add", pkg, "--verbose"], { cwd });

		// With verbose output on, bun prints one line per installed package
		// (e.g., "foo@1.0.0"), including all installed dependencies.
		// These lines are between the two delimiting lines found here:
		// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
		const lines = stdout.split("\n");
		const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
		const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
		const installedPackages = lines.slice(beginHash + 1, endHash);
		return installedPackages;
	}
}
