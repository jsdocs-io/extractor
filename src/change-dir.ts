import { Result } from "neverthrow";
import { OsError } from "./errors";

export const changeDir = Result.fromThrowable(
  process.chdir,
  (e) => new OsError("failed to change directory", { cause: e }),
);
