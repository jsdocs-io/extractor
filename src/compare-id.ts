export type WithId = {
  id: string;
};

export const compareId = (a: WithId, b: WithId): number =>
  a.id.localeCompare(b.id);
