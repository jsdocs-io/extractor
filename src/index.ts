export {
  extractDeclarations,
  type ExtractDeclarationsOptions,
  type ExtractedDeclaration,
  type ExtractedDeclarationKind,
} from "./extract-declarations";
