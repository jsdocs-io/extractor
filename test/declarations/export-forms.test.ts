import dedent from "ts-dedent";
import {
  ModuleKind,
  ModuleResolutionKind,
  Project,
  ScriptTarget,
} from "ts-morph";
import { expect, test } from "vitest";
import { extractDeclarations } from "../../src";

// Various kinds of `export` declarations and forms.
// See https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export.

test("export var list", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export var foo: string, bar: number;
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export let list", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export let foo: string, bar: number;
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export const assignment list", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export const foo = 1, bar = 2;
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export let assignment list", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export let foo = 1, bar = 2;
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export var assignment list", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export var foo = 1, bar = 2;
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export named function", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export function foo() {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export named class", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export class Foo() {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export named generator function", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export function* generatorFoo() {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

// TODO:
test("export const object destructure", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export const { foo, bar: baz } = { foo: 1, bar: 2 };
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

// TODO:
test("export const array destructure", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export const [foo, bar] = [1, 2];
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export list", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    function foo() {}
    const bar = "";
    export { foo, bar };
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export list as renamed", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    function foo() {}
    const bar = "";
    export { foo as myFoo, bar as myBar };
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

// Currently not supported by TypeScript.
// See: https://github.com/microsoft/TypeScript/issues/40594.
test("export list as renamed to string", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    function foo() {}
    const bar = "";
    export { foo as "my foo", bar as "my bar" };
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export list as renamed to default", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    function foo() {}
    const bar = "";
    export { foo as default, bar };
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default expression", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export default 1 + 1;
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default named function", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export default function foo() {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default named class", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export default class Foo() {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default named generator function", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export default function* generatorFoo() {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default anonymous function", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export default function() {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default anonymous class", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export default class {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default anonymous generator function", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export default function* () {}
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export all from module", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  project.createSourceFile(
    "module.ts",
    dedent`
    export const foo: string;
    `,
  );
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export * from "./module";
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export all from module as renamed", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  project.createSourceFile(
    "module.ts",
    dedent`
    export const foo: string;
    `,
  );
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export * as myModule from "./module";
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export list from module", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  project.createSourceFile(
    "module.ts",
    dedent`
    export function foo() {}
    export const bar: string;
    export const baz: number;
    `,
  );
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export { foo, bar } from "./module";
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export list as renamed from module", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  project.createSourceFile(
    "module.ts",
    dedent`
    export function foo() {}
    export const bar: string;
    export const baz: number;
    `,
  );
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export { foo as myFoo, bar as myBar } from "./module";
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default from module", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  project.createSourceFile(
    "module.ts",
    dedent`
    export default function foo() {}
    export const bar: string;
    export const baz: number;
    `,
  );
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export { default } from "./module";
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export default as renamed from module", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  project.createSourceFile(
    "module.ts",
    dedent`
    export default function foo() {}
    export const bar: string;
    export const baz: number;
    `,
  );
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export { default as myDefault } from "./module";
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});

test("export empty", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.ts",
    dedent`
    export {};
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});
