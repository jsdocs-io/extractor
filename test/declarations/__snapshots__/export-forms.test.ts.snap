// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`export all from module 1`] = `
[
  {
    "docs": [],
    "file": "/module.ts",
    "id": "variable.foo",
    "kind": "variable",
    "line": 1,
    "name": "foo",
    "signature": "const foo: string;",
  },
]
`;

exports[`export all from module as renamed 1`] = `
[
  {
    "declarations": [
      {
        "docs": [],
        "file": "/module.ts",
        "id": "namespace.myModule.variable.foo",
        "kind": "variable",
        "line": 1,
        "name": "foo",
        "signature": "const foo: string;",
      },
    ],
    "docs": [],
    "file": "/module.ts",
    "id": "namespace.myModule",
    "kind": "namespace",
    "line": 1,
    "name": "myModule",
    "signature": "module "module.ts" {}",
  },
]
`;

exports[`export const array destructure 1`] = `[]`;

exports[`export const assignment list 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 1,
    "name": "bar",
    "signature": "const bar: number;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.foo",
    "kind": "variable",
    "line": 1,
    "name": "foo",
    "signature": "const foo: number;",
  },
]
`;

exports[`export const object destructure 1`] = `[]`;

exports[`export default anonymous class 1`] = `
[
  {
    "constructors": [],
    "docs": [],
    "file": "/index.ts",
    "id": "class.default",
    "kind": "class",
    "line": 1,
    "methods": [],
    "name": "default",
    "properties": [],
    "signature": "class {}",
  },
]
`;

exports[`export default anonymous function 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.default",
    "kind": "function",
    "line": 1,
    "name": "default",
    "signature": "default: () => void;",
  },
]
`;

exports[`export default anonymous generator function 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.default",
    "kind": "function",
    "line": 1,
    "name": "default",
    "signature": "default: () => Generator<never, void, unknown>;",
  },
]
`;

exports[`export default as renamed from module 1`] = `
[
  {
    "docs": [],
    "file": "/module.ts",
    "id": "function.myDefault",
    "kind": "function",
    "line": 1,
    "name": "myDefault",
    "signature": "myDefault: () => void;",
  },
]
`;

exports[`export default expression 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.default",
    "kind": "variable",
    "line": 1,
    "name": "default",
    "signature": "const default: number;",
  },
]
`;

exports[`export default from module 1`] = `
[
  {
    "docs": [],
    "file": "/module.ts",
    "id": "function.default",
    "kind": "function",
    "line": 1,
    "name": "default",
    "signature": "default: () => void;",
  },
]
`;

exports[`export default named class 1`] = `
[
  {
    "constructors": [],
    "docs": [],
    "file": "/index.ts",
    "id": "class.default",
    "kind": "class",
    "line": 1,
    "methods": [],
    "name": "default",
    "properties": [],
    "signature": "class Foo {}",
  },
]
`;

exports[`export default named function 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.default",
    "kind": "function",
    "line": 1,
    "name": "default",
    "signature": "default: () => void;",
  },
]
`;

exports[`export default named generator function 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.default",
    "kind": "function",
    "line": 1,
    "name": "default",
    "signature": "default: () => Generator<never, void, unknown>;",
  },
]
`;

exports[`export empty 1`] = `[]`;

exports[`export let assignment list 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 1,
    "name": "bar",
    "signature": "let bar: number;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.foo",
    "kind": "variable",
    "line": 1,
    "name": "foo",
    "signature": "let foo: number;",
  },
]
`;

exports[`export let list 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 1,
    "name": "bar",
    "signature": "let bar: number;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.foo",
    "kind": "variable",
    "line": 1,
    "name": "foo",
    "signature": "let foo: string;",
  },
]
`;

exports[`export list 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.foo",
    "kind": "function",
    "line": 1,
    "name": "foo",
    "signature": "foo: () => void;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 2,
    "name": "bar",
    "signature": "const bar: string;",
  },
]
`;

exports[`export list as renamed 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.myFoo",
    "kind": "function",
    "line": 1,
    "name": "myFoo",
    "signature": "myFoo: () => void;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.myBar",
    "kind": "variable",
    "line": 2,
    "name": "myBar",
    "signature": "const myBar: string;",
  },
]
`;

exports[`export list as renamed from module 1`] = `
[
  {
    "docs": [],
    "file": "/module.ts",
    "id": "function.myFoo",
    "kind": "function",
    "line": 1,
    "name": "myFoo",
    "signature": "myFoo: () => void;",
  },
  {
    "docs": [],
    "file": "/module.ts",
    "id": "variable.myBar",
    "kind": "variable",
    "line": 2,
    "name": "myBar",
    "signature": "const myBar: string;",
  },
]
`;

exports[`export list as renamed to default 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.default",
    "kind": "function",
    "line": 1,
    "name": "default",
    "signature": "default: () => void;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 2,
    "name": "bar",
    "signature": "const bar: string;",
  },
]
`;

exports[`export list as renamed to string 1`] = `[]`;

exports[`export list from module 1`] = `
[
  {
    "docs": [],
    "file": "/module.ts",
    "id": "function.foo",
    "kind": "function",
    "line": 1,
    "name": "foo",
    "signature": "foo: () => void;",
  },
  {
    "docs": [],
    "file": "/module.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 2,
    "name": "bar",
    "signature": "const bar: string;",
  },
]
`;

exports[`export named class 1`] = `
[
  {
    "constructors": [],
    "docs": [],
    "file": "/index.ts",
    "id": "class.Foo",
    "kind": "class",
    "line": 1,
    "methods": [],
    "name": "Foo",
    "properties": [],
    "signature": "class Foo {}",
  },
]
`;

exports[`export named function 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.foo",
    "kind": "function",
    "line": 1,
    "name": "foo",
    "signature": "foo: () => void;",
  },
]
`;

exports[`export named generator function 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "function.generatorFoo",
    "kind": "function",
    "line": 1,
    "name": "generatorFoo",
    "signature": "generatorFoo: () => Generator<never, void, unknown>;",
  },
]
`;

exports[`export var assignment list 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 1,
    "name": "bar",
    "signature": "var bar: number;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.foo",
    "kind": "variable",
    "line": 1,
    "name": "foo",
    "signature": "var foo: number;",
  },
]
`;

exports[`export var list 1`] = `
[
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.bar",
    "kind": "variable",
    "line": 1,
    "name": "bar",
    "signature": "var bar: number;",
  },
  {
    "docs": [],
    "file": "/index.ts",
    "id": "variable.foo",
    "kind": "variable",
    "line": 1,
    "name": "foo",
    "signature": "var foo: string;",
  },
]
`;
